### DISCUSSION TITLE
Deprecating product types in favor of options/traits

### DISCUSSION THREAD

**Discussion by joshuatf** [woocommerce/woocommerce#44814] [General]
In an effort to simplify user experience with products by combining simple and variable products into a single template (pcQJnU-2f6-p2), we accidentally unearthed a series of long-standing issues within products in WooCommerce.

After looking through a number of extensions that make use of these features and heavy discussion around templating compatibility with types and templates, I've been contemplating if it's possible to completely remove product types in favor of product traits. At the very least, we need a better, more formal trait management system.

Note that in this post I use "trait" or "product trait" to refer to the similar existing concept of "product type options" in order to differentiate this idea from product types.

## Difference between product types and traits

Product types and traits are tools to help control the features of a kind of product, but are commonly misused in extensions.

### Product traits

<img width="312" alt="Screenshot 2024-02-20 at 11 43 55 AM" src="https://github.com/woocommerce/woocommerce/assets/10561050/14aaeff8-c9f3-4b25-8037-0cd99e931715">

Traits extend the functionality of existing product types and can be applied to some or all product types. A trait can be compatible with another trait

Traits are a very flexible way to add in functionality to existing product types and features, that provides the opportunity for cross-compatibility with other traits going forward.

### Product types
<img width="385" alt="Screenshot 2024-02-19 at 6 44 04 PM" src="https://github.com/woocommerce/woocommerce/assets/10561050/6372cb26-3c25-46e8-9092-3c0d8b089c5e">

Product types are a way to distinguish the behavior of a specific kind of product and explicitly declare that a product is not compatible with other product types.

In many ways, product types are really just like product traits with the exception that they cannot ever be compatible with another product type.

### Types across templates

This is what spurred this exploration and relevant discussion around types/traits/templates (pcQJnU-2f6-p2).  The new product editing experience attempts to simplify merchant experience by allowing creation of simple or variable products from within the same product template.  It attempts to reduce the complexity of types that is usually offloaded to merchants by creating a simpler UI to create the intended product.

The issue with attempting to incorporate multiple product types within a single template is that we create a "tug-of-war" between two or more product types when an underlying action triggers a product type change.

While this may be something we can solve in the simple product template for core product types, this problem gets further compounded when extensions add their own templates.  For example, if we create a "Gift card" template, **this template is now forced to figure out which other traits its compatible with, which product types its compatible with, and how product types should be converted to one another.**

It's possible that this means we should just stick to a 1:1 for product templates and product types, but I do see high value in the opportunity for templates to be used as starting points for a variety of combinations of product types, traits, and predefined product data.

### Inability to share functionality

Product types make a definitive assertion that they will never be compatible with other product types since a product can never have multiple types.

The fact that a type declares incompatibility with other types isn't inherently a bad thing, but unfortunately this is very limiting as we've seen time and time again as we experience lock-in with a product type that we later realize could benefit from integrating with another product type's functionality.

A good example of this is WooCommerce Subscriptions which creates 2 new product types:

* `subscription`
* `variable-subscription`

It's easy to say that this should have been a trait as multiple product types could be "subscribable," but given that subscriptions needs both `simple` and `variable` product types, it really did not have much choice other than creating 2 new brand new product types. So instead of having a single product type that can have a combination of being subscribable and/or variable, we end up with 4 product types in total to maintain (`simple, `variable`, `subscription`, `variable-subscription`).

I think the argument could easily be made that product types are fine as is and what we're really seeing here is misuse of product types. **But in reality it is very difficult, if not impossible, to know in advance that a newly created product type will never be compatible with any other product types.**

## Why aren't product traits used more often?

For most of the extensions and use cases that I've come across, traits seem to be the more correct answer to the problem being solved. They can extend functionality across the entire product or limit it to certain areas where not needed or incompatible.

So why do more extensions seem to prefer using product types over traits?

* **Lack of official API** - currently we don't have any actual API to support the idea of traits. They are purely conceptual with the exception of a single filter product_type_options that allows adding a trait/option.
* **Compatibility** - This is controlled by CSS class such as show_if_x or hide_if_x and also does not have any official system.
* **No clear guidelines** - We don't have very good guidelines or documentation at the moment for when to create a new product type or when a trait should be used.
* **Surfacing the new feature** - A product type seems much more prevalent in the previous experience and extensions wanting to showcase this feature may opt to create a product type instead of a less obvious trait.
* **Separation of behavior** - The lack of conflicts with other plugins and ability to define specific behavior, either around checkout, button text, or any other product related feature is much easier when you can be sure you're not contending with other plugins.

The above problems can all be resolved between a new compatibility system for traits, the introduction of product templates, and better documentation around how to use these features. The one exception to this is dealing with the "separation of behavior" issue or conflict resolution between two traits modifying similar behavior.

## Proof of concept

In https://github.com/woocommerce/woocommerce/pull/44585, I've introduced the concept of compatibility between various traits. What this means is that traits can choose to be:

* Compatible with only the default simple WooCommerce product traits
* Compatible with some or all other traits
* Not compatible with any other traits

This effectively covers everything product types can do and more, but comes with the major advantage that compatibility between traits can be declared at a future date.

https://github.com/woocommerce/woocommerce/pull/44657 (while incomplete and very hacky with its heavy reliance on the __call magic method) makes an effort to continue supporting $product->get_type() and allowing product traits to continue extending the product class with custom methods. If we moved towards this solution, we would want to come up with an extensive deprecation strategy to phase out types while continuing to support them since so many extensions rely on plugin types and even product types registered by other third party extensions.

Also something that the POC hints at is the idea that **many core features, even down to images and downloads, could be represented by traits to allow easier declaration of compatibility by custom traits if they do not support these base features**. This better encapsulates logic around these features, but also more easily allows us to hide or disable features that are not compatible with third party traits.

## How are conflicts dealt with?

The unresolved issue with moving towards a trait-based system is dealing with collision between 2 product types. As a simple example, let's take the `add_to_cart_text()` method in product types that is used to determine which "add to cart text" to show on the product page.

Moving towards a trait-based system means that two different traits may be trying to modify this text simultaneously and fighting for control. However, I think this _might be_ less of an issue than it appears on the surface.

In concept, traits should not compatible with one another until declaring compatibility. This means that luckily there is an opportunity to resolve conflicts since compatibility is explicitly declared. If we rely on the filter that already exists for "add to cart text" and continuing with our example from earlier:

```php
class WC_Product_Trait_Subscribable extends WC_Product_Trait {
    public function __construct() {
    // Note that this is higher than the default priority of 10 since presumably core traits would run on 10.
      add_filter( 'woocommerce_product_add_to_cart_text', array( $this, 'filter_add_to_cart_text' ), 20, 2 );
    }

    public function filter_add_to_cart_text( $text, $product ) {
        if ( $product->has_trait( 'subscribable' ) ) {
            // Note that these strings are not accurate to the extension and are purely to demonstrate conflict resolution with cross-compatible traits.
            return $product->has_trait( 'variable' ) ? 'Choose subscription' : 'Sign up now';
        }

        return $text;
    }

    public static function get_compatible_traits() {
        return array_merge(
            array_keys( WC()->product_traits()->get_simple_product_traits() ),
            array(
                'variable'
            )
        );
    }
}
```

## Next steps

I think that creating a product trait system to check compatibility is an easy win, but deprecating product types is a much larger project. This is obviously a huge change and not something we should consider without extensive discussion.

I think in order for this to be a viable solution, we need to validate a few points:

2. [ ] We can deprecate product types without breaking backwards compatibility and can continue to support product types for at least 2 years.
1. [ ] The conversion of types to traits and reusability of functionality across products/templates makes this endeavor worthwhile.
3. [ ] Product traits can cover all existing cases that product type covers.
4. [ ] Conflict resolution does not pose a major problem for cross-compatible traits.

I realize this is a potentially controversial take on product types and definitely welcome any criticism or concerns around this approach.  While I really like this concept in theory, I'm sure there are a number of cases that will make this transition very challenging to accomplish.

@woocommerce/mothra @woocommerce/somewherewarm  @woocommerce/rubik @woocommerce/woo-fse

**Comment by helgatheviking** [comment]
Just subscribing myself to this since we were discussing it in the community slack.

**Comment by BFTrick** [comment]
👍 to product traits.

It gives shop managers much more flexibility into how they create their products and it's far less rigid.

**Comment by seb86** [comment]
This actually all makes sense and your already using a smaller similar system for payment gateways identifying in what they support. A good example of this trait method could be tested with the extension "All Products for Subscriptions".

**Comment by mocassinis** [comment]
I am having quite many good and bad thoughts on traits. And foremost how we can keep compatibility. 

But on then other hand this line was quite interesting!! ...a product type that we later realize could benefit from integrating with another product type's functionality.

I am on traits!

**Comment by skunkbad** [comment]
As the web dev for a fairly large store with lots of products and custom features, I'm imagining the proposed changes being a nightmare for us. If there isn't anything seriously wrong with the way this all works now, I vote for leaving things alone.

**Reply by joshuatf** [reply]
>If there isn't anything seriously wrong with the way this all works now

There are issues with how it functions right now which spurred this discussion and I'll try to update this post next week to provide more context around those discussions.

The short of that discussion is that the current data structure of product types limits design opportunities and weakens the user experience.  While attempts have been made to work around this, they've all surfaced edge cases with plugins that rely heavily on product types.

>I'm imagining the proposed changes being a nightmare for us.

Assuming that we find a way to make this backwards compatible (which would be a necessity before we begin on this project) and all existing extension code continues to work as is, what are your top concerns around this change that would make it challenging for you?

**Comment by Drivingralle** [comment]
I vote for moving to traits.
Requiring the new interface to work with them, would make the change teaching and developing for it clear. 

A solid way to support old product types is important.

**Comment by simplysaru** [comment]
I vote for moving to traits as it seems more flexible and a versatile solution.

**Comment by jonathan-dejong** [comment]
I like this idea. 
It's always been a strange experience where products and product variations are two separate post types but essentially the same except with some certain fields.. forcing us to work around things like not having image galleries for product variations for example. 

In my mind this also means we can start moving away from separate product type templates which all calls slightly different hooks or the same hooks but in different locations. As an example adding something above the add to cart button on a single product template requires separate hooks depending on the product type 🤪 

---

Something that popped up in my head: given that WP core moves towards json based configs like theme.json and block.json does it make sense to declare trait supports etc. as a json file? 
Trait name, trait compatibility etc. 
It may be helpful bridging the backend (PHP) and frontend (new block based editor)? 

---

On the note of "Trait" is it a confusing name? `Product component` or `Product Capability` ? 
I mean in the POC we're using abstract classes and calling them traits which feels a bit wrong 😄 

---

Backwards compatibility is of course important and part of WCs guidelines, but am I the only one getting a bit tired of us constantly tying a rope around our hands preventing progress in the name of infinitely supporting old solutions? 
_You have to crack a few eggs to make an omelette_

Because I know that this may stir some emotions I'm saying this as a person in an agency building a lot of WC based projects as well as our own e-commerce products on top of WC. We'd be very affected by breaking backwards compatibility but if not doing it comes at the cost of continuously being passed by other platforms it's just a slow walk towards irrelevance.
